ldap/servers/slapd/connection.c

```diff
@@ -425,71 +419,191 @@ connection_reset(Connection *conn, int ns, PRNetAddr *from, int fromLen __attrib
     conn->c_serveripaddr = slapi_ch_strdup(str_destip);
 }
 
-/* Create a pool of threads for handling the operations */
+/*
+ * worker thread context destructor.
+ */
+void
+op_thread_tinfo_destroy(pc_tinfo_t *tinfo)
+{
+    /* Removing context from snmp slot */
+    pthread_mutex_lock(&g_pc.snmp.mutex);
+    if (tinfo->idx <= g_pc.snmp.nbthreads &&
+        g_pc.snmp.threads[tinfo->idx] == tinfo) {
+        g_pc.snmp.threads[tinfo->idx] = NULL;
+    }
+    pthread_mutex_unlock(&g_pc.snmp.mutex);
+    /* Unqueing context from waiting/busy queues */
+    pthread_mutex_lock(&g_pc.mutex);
+    ll_unlink(&tinfo->q);
+    pthread_mutex_unlock(&g_pc.mutex);
+    /* Clear the snmp context */
+    snmp_thread_counters_cleanup(&tinfo->snmp_vars);
+    /* Free the operation */
+    operation_free(&tinfo->op, NULL);
+    /* And destroy the context */
+    slapi_ch_free((void**)&tinfo);
+}
+
+/*
+ * Alloc and initialize a worker thread context
+ * The caller should hold g_pc.snmp.mutex
+ */
+pc_tinfo_t *
+op_thread_tinfo_init(int thread_idx)
+{
+    pc_tinfo_t *tinfo = (pc_tinfo_t*) slapi_ch_calloc(1, sizeof(*tinfo));
+    int rc = 0;
+
+    ll_init(&tinfo->q, tinfo);
+    tinfo->idx = thread_idx;
+    snmp_thread_counters_init(&tinfo->snmp_vars);
+    tinfo->op = operation_new(plugin_build_operation_action_bitmap(0, plugin_get_server_plg()));
+    if (!tinfo->op) {
+        slapi_log_err(SLAPI_LOG_ERR, "op_thread_tinfo_init",
+                      "Failed to create worker thread [%d] operation.\n", thread_idx);
+    } else if ((rc = pthread_mutex_init(&tinfo->mutex, NULL))) {
+        slapi_log_err(SLAPI_LOG_ERR, "op_thread_tinfo_init",
+                      "Failed to initialize worker thread [%d] lock.\n", thread_idx);
+    } else if ((rc = pthread_cond_init(&tinfo->cv, NULL))) {
+        slapi_log_err(SLAPI_LOG_ERR, "op_thread_tinfo_init",
+                      "Failed to initialize worker thread [%d] condition variable.\n", thread_idx);
+    } else {
+        return tinfo;
+    }
+    op_thread_tinfo_destroy(tinfo);
+    return NULL;
+}
+
+/*
+ * Increase or decrease the number of worker threads
+ */
+void
+op_thread_set_threads_number(int threadsnumber)
+{
+    int oldnbthreads = 0;
```

Comment:
If op_thread_set_thread_number is called in parallel, should not serialize the call ?

Reply:
Good catch. I should add a specific mutex to serialize this function

Reply:
It is now fixed by latest commit